feat: add loggers package

emmenko opened the pull request:
Partially part of #1444 

This PR open sources our internal loggers package, used by our backend services.

tdeekens commented on packages-backend/loggers/src/custom-formats/json.ts:
```diff
@@ -0,0 +1,55 @@
+function replacer(key: string, value: Buffer | string) {
+  if (key === 'queryJsonString') {
+    // We need to stringify the value as otherwise Kibana cardinality explodes.
+    // Ref: https://commercetools.slack.com/archives/C011FLZ1JLW/p1588066334005900
```
I think we should remove references from internal Slack convos.

tdeekens commented on packages-backend/loggers/src/create-error-report-logger.ts:
```diff
@@ -0,0 +1,89 @@
+
+function createErrorReportLogger(options: LoggerOptions = {}) {
+  // Sentry
+  if (options.sentry) {
```
Hm. Personally I find it rather specific that we use Sentry. Do we know if others using this package would too or should we allow for more reporting targets?

emmenko replied:
In my defense, the package says "_Opinionated JSON loggers_" 😜

emmenko commented:
I understand your points, let's have a chat then 🙂

tdeekens commented on packages-backend/loggers/src/custom-formats/json.ts:
```diff
@@ -0,0 +1,54 @@
+  delete info.level;
+
+  // Replace / rename fields
+  replaceField(info, 'meta.req.query', 'meta.req.queryJsonString');
```
Should be an option `omitMetaPaths: String[]` I think.

emmenko replied:
Yeah, probably we need to provide some kind of fields mapping.

```js
{
  rewriteMetaInfoFields: {
    'meta.req.query': 'meta.req.queryJsonString'
  }
}
```

An alternative might be to allow to pass a custom `logform` formatter. We can also expose a bunch of default formatter (e.g. json, kibana, etc). The kibana formatter can also be additionally configured with the fields rewriter.

tdeekens replied:
Yeah. I think sane defaults which can be overwritten would be good. Maybe also by just saying: I want a Kibana logger as a type argument you get those defaults.

tdeekens commented on packages-backend/loggers/src/create-access-logger.ts:
```diff
@@ -0,0 +1,51 @@
+  return req[propName];
+};
+
+const createAccessLogger = (options: LoggerOptions = {}) => {
```
This is a middleware I think. So to have it inline with others maybe it's `createAccessLoggerMiddleware(options)`.

tdeekens commented on packages-backend/loggers/src/env.ts:
```diff
@@ -0,0 +1,6 @@
+const isProd = env === 'production';
+const isTest = env === 'test';
+
+export { isDev, isProd, isTest };
```
I would also have this as an arg to the middlewars. `envGetters` maybe. Not 100% also just making notes for myself.

emmenko commented on packages-backend/loggers/src/create-error-report-logger.ts:
```diff
@@ -0,0 +1,89 @@
+
+import * as Sentry from '@sentry/node';
+import * as env from './env';
+import redactInsecureRequestHeaders from './utils/redact-insecure-request-headers';
```
PS: one of the intents with this loggers package was to provide preconfigured loggers, meaning that for common cases there should be no need to provide extra configuration.

We could maybe have a default list of "insecure" headers, while still allowing to extend that list.

emmenko commented:
I cleaned things up and remove all unnecessary things. I also removed Sentry, as I think we can keep this internal for now.

@tdeekens can you take another look?

tdeekens commented on packages-backend/loggers/src/index.ts:
```diff
@@ -0,0 +1,12 @@
+
+// Re-export winston for convenience
+import * as winston from 'winston';
+export { winston };
```
Nit: doesn't TS allow `export * as winston` from 'winston'`?
